File: app/src/main/java/com/amaury/pointage/v2/CompanyIncomeTaxRateStoreV2.kt
package com.amaury.pointage.v2

import android.content.Context
import com.amaury.pointage.SalaryCompanyStore
import com.amaury.pointage.v2.engine.CompanyIncomeTaxRateResolverV2
import org.json.JSONArray
import org.json.JSONObject
import java.time.YearMonth

/** Stockage local et daté du taux personnel de prélèvement à la source pour une entreprise. */
object CompanyIncomeTaxRateStoreV2 {
    private const val KEY = "income_tax_rates_v2"
    private const val STORAGE_WARNING =
        "PAS : stockage local des taux datés incohérent ; calcul après impôt bloqué et ancien taux non réutilisé."

    data class ReadResult(
        val records: List<CompanyIncomeTaxRateResolverV2.Record>,
        val reliable: Boolean,
        val warnings: List<String>
    )

    internal fun companyUnavailableResult(): ReadResult = ReadResult(
        records = emptyList(),
        reliable = false,
        warnings = listOf("PAS : entreprise absente ou stockage des entreprises non fiable ; taux personnel inaccessible.")
    )

    fun read(context: Context, companyId: String): ReadResult {
        if (companyId.isBlank()) return companyUnavailableResult()
        return SalaryCompanyStore.withConfirmedCompany(context, companyId) {
            readConfirmed(context, companyId)
        } ?: companyUnavailableResult()
    }

    fun list(context: Context, companyId: String): List<CompanyIncomeTaxRateResolverV2.Record> =
        read(context, companyId).records

    fun save(context: Context, companyId: String, record: CompanyIncomeTaxRateResolverV2.Record): Boolean {
        if (companyId.isBlank() || !valid(record)) return false
        return SalaryCompanyStore.withConfirmedCompany(context, companyId) {
            val stored = readConfirmed(context, companyId)
            if (!stored.reliable) return@withConfirmedCompany false
            val items = stored.records.toMutableList()
            val index = items.indexOfFirst { it.id == record.id }
            if (index >= 0) items[index] = record else items += record
            writeConfirmed(context, companyId, items)
        } == true
    }

    fun remove(context: Context, companyId: String, id: String): Boolean {
        if (companyId.isBlank() || id.isBlank()) return false
        return SalaryCompanyStore.withConfirmedCompany(context, companyId) {
            val stored = readConfirmed(context, companyId)
            if (!stored.reliable) return@withConfirmedCompany false
            writeConfirmed(context, companyId, stored.records.filterNot { it.id == id })
        } == true
    }

    fun resolve(context: Context, companyId: String, period: YearMonth): CompanyIncomeTaxRateResolverV2.Snapshot {
        if (companyId.isBlank()) return resolve(companyUnavailableResult(), period)
        return SalaryCompanyStore.withConfirmedCompany(context, companyId) {
            resolve(readConfirmed(context, companyId), period)
        } ?: resolve(companyUnavailableResult(), period)
    }

    /**
     * hasDatedRecords=true est volontaire en cas de corruption : le fallback legacy ne doit
     * jamais masquer un store V2 incohérent en réutilisant un ancien taux sans période.
     */
    internal fun resolve(stored: ReadResult, period: YearMonth): CompanyIncomeTaxRateResolverV2.Snapshot {
        if (stored.reliable) return CompanyIncomeTaxRateResolverV2.resolve(stored.records, period)
        return CompanyIncomeTaxRateResolverV2.Snapshot(
            rate = null,
            ratePercent = null,
            source = null,
            hasDatedRecords = true,
            reliable = false,
            warnings = stored.warnings.ifEmpty { listOf(STORAGE_WARNING) }
        )
    }

    internal fun decode(raw: String): ReadResult = runCatching {
        val array = JSONArray(raw)
        val records = mutableListOf<CompanyIncomeTaxRateResolverV2.Record>()
        var malformed = false
        for (i in 0 until array.length()) {
            val record = fromJson(array.opt(i) as? JSONObject)
            if (record == null) malformed = true else records += record
        }
        if (records.groupingBy { it.id }.eachCount().any { it.value > 1 }) malformed = true
        ReadResult(
            records = records,
            reliable = !malformed,
            warnings = if (malformed) listOf(STORAGE_WARNING) else emptyList()
        )
    }.getOrElse {
        ReadResult(emptyList(), false, listOf(STORAGE_WARNING))
    }

    private fun readConfirmed(context: Context, companyId: String): ReadResult {
        val raw = SalaryCompanyStore.prefs(context, companyId).getString(KEY, "[]")
            ?: return ReadResult(emptyList(), false, listOf(STORAGE_WARNING))
        return decode(raw)
    }

    private fun writeConfirmed(
        context: Context,
        companyId: String,
        items: List<CompanyIncomeTaxRateResolverV2.Record>
    ): Boolean {
        if (items.any { !valid(it) } || items.groupingBy { it.id }.eachCount().any { it.value > 1 }) return false
        val array = JSONArray()
        items.forEach { array.put(toJson(it)) }
        return SalaryCompanyStore.prefs(context, companyId)
            .edit()
            .putString(KEY, array.toString())
            .commit()
    }

    private fun toJson(record: CompanyIncomeTaxRateResolverV2.Record) = JSONObject()
        .put("id", record.id)
        .put("ratePercent", record.ratePercent)
        .put("effectiveFrom", record.effectiveFrom?.toString() ?: JSONObject.NULL)
        .put("effectiveTo", record.effectiveTo?.toString() ?: JSONObject.NULL)
        .put("source", record.source ?: JSONObject.NULL)

    private fun fromJson(o: JSONObject?): CompanyIncomeTaxRateResolverV2.Record? {
        o ?: return null
        val id = (o.opt("id") as? String)?.trim()?.takeIf { it.isNotBlank() } ?: return null
        val ratePercent = (o.opt("ratePercent") as? Number)?.toDouble() ?: return null
        val effectiveFrom = parseRequiredMonth(o, "effectiveFrom") ?: return null
        val effectiveTo = parseOptionalMonth(o, "effectiveTo") ?: return null
        val source = (o.opt("source") as? String)?.trim()?.takeIf { it.isNotBlank() } ?: return null
        return CompanyIncomeTaxRateResolverV2.Record(
            id = id,
            ratePercent = ratePercent,
            effectiveFrom = effectiveFrom,
            effectiveTo = effectiveTo.value,
            source = source
        ).takeIf(::valid)
    }

    private data class OptionalMonth(val value: YearMonth?)

    private fun parseRequiredMonth(o: JSONObject, key: String): YearMonth? {
        val raw = o.opt(key) as? String ?: return null
        if (raw.isBlank() || raw == "null") return null
        return runCatching { YearMonth.parse(raw) }.getOrNull()
    }

    private fun parseOptionalMonth(o: JSONObject, key: String): OptionalMonth? = when (val raw = o.opt(key)) {
        null, JSONObject.NULL -> OptionalMonth(null)
        is String -> {
            if (raw.isBlank() || raw == "null") OptionalMonth(null)
            else runCatching { OptionalMonth(YearMonth.parse(raw)) }.getOrNull()
        }
        else -> null
    }

    private fun valid(record: CompanyIncomeTaxRateResolverV2.Record): Boolean {
        if (record.id.isBlank()) return false
        if (!record.ratePercent.isFinite() || record.ratePercent < 0.0 || record.ratePercent > 100.0) return false
        val start = record.effectiveFrom ?: return false
        val end = record.effectiveTo
        if (end != null && end < start) return false
        return !record.source.isNullOrBlank()
    }
}

File: ios/HPTravail/HPTravail/SalaryV2/CompanyIncomeTaxRateStoreV2.swift
import Foundation

/// Stockage local des taux PAS confirmés par mois pour l'écran Salaire V2 iOS.
/// Un taux n'est jamais appliqué hors de sa période explicite.
final class CompanyIncomeTaxRateStoreV2 {
    private struct StoredRecord: Codable {
        let id: String
        let ratePercent: Double
        let from: String
        let to: String?
        let source: String
    }

    private let defaults: UserDefaults
    private func key(companyId: String) -> String {
        "income_tax_rates_v2." + companyId.trimmingCharacters(in: .whitespacesAndNewlines)
    }

    init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
    }

    func snapshot(companyId: String, for period: YearMonthV2) -> CompanyIncomeTaxRateResolverV2.Snapshot {
        let storageKey = key(companyId: companyId)
        guard !companyId.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            return .init(rate: nil, ratePercent: nil, source: nil, hasDatedRecords: false, reliable: false, warnings: ["PAS : entreprise non confirmée ; calcul après impôt bloqué."])
        }
        guard let data = defaults.data(forKey: storageKey) else {
            return CompanyIncomeTaxRateResolverV2.resolve(records: [], period: period)
        }
        do {
            let stored = try JSONDecoder().decode([StoredRecord].self, from: data)
            let records: [CompanyIncomeTaxRateResolverV2.Record] = try stored.map {
                guard let from = YearMonthV2($0.from) else { throw StoreError.invalidPeriod }
                let to: YearMonthV2?
                if let rawTo = $0.to {
                    guard let parsed = YearMonthV2(rawTo) else { throw StoreError.invalidPeriod }
                    to = parsed
                } else {
                    to = nil
                }
                return CompanyIncomeTaxRateResolverV2.Record(
                    id: $0.id,
                    ratePercent: $0.ratePercent,
                    effectiveFrom: from,
                    effectiveTo: to,
                    source: $0.source
                )
            }
            return CompanyIncomeTaxRateResolverV2.resolve(records: records, period: period)
        } catch {
            return .init(
                rate: nil,
                ratePercent: nil,
                source: nil,
                hasDatedRecords: true,
                reliable: false,
                warnings: ["PAS : stockage local incohérent ; calcul après impôt bloqué."]
            )
        }
    }

    private enum StoreError: Error { case invalidPeriod }

    @discardableResult
    func confirm(companyId: String, ratePercent: Double, period: YearMonthV2, source: String) -> Bool {
        let storageKey = key(companyId: companyId)
        guard !companyId.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else { return false }
        let trimmed = source.trimmingCharacters(in: .whitespacesAndNewlines)
        guard ratePercent.isFinite, (0...100).contains(ratePercent), !trimmed.isEmpty else { return false }

        var records: [StoredRecord] = []
        if let data = defaults.data(forKey: storageKey),
           let decoded = try? JSONDecoder().decode([StoredRecord].self, from: data) {
            records = decoded.filter { $0.from != period.description }
        } else if defaults.object(forKey: storageKey) != nil {
            return false
        }

        records.append(.init(
            id: "pas-\(period.description)",
            ratePercent: ratePercent,
            from: period.description,
            to: period.description,
            source: trimmed
        ))
        guard let data = try? JSONEncoder().encode(records) else { return false }
        defaults.set(data, forKey: storageKey)
        return true
    }
}
